Grammar !

johangirod opened the pull request:
- extends dottedName possibility
- extends unit possibility
- add expression (bool and arithmetic)

@EmileRolley : première itération ici. Je bloque sur le fait de ne pas permettre les mots-clés « oui » et « non »  comme identifier. Et la prochaine étape : utiliser un [scanner externe](https://tree-sitter.github.io/tree-sitter/creating-parsers#external-scanners) pour l'indentation.

johangirod committed:
Improve grammar

- extends dottedName possibility
- extends unit possibility

johangirod committed:
Ajoute les expressions arithmétiques

mquandalle commented on grammar.js:
```diff
@@ -92,7 +147,19 @@ module.exports = grammar({
-        unit: ($) => seq(token.immediate(unit), optional($.exposant)),
-        exposant: () => token.immediate(choice("²", "³")),
+        unit: ($) => seq(unit, optional($.exposant)),
+        exposant: () => /²|³/,
```
Je me balade ici, les autres exposants existent aussi : ⁰, ¹, ⁵, etc. Les codes unicodes sont de U+2070 à U+2079:

```suggestion
        exposant: () => /[\u{2070}-\u{2079}]+/u,
```

johangirod replied:
Haha merci :p
On pourrait faire une petite séance de pair un de ces quatre si le sujet t'interesse !

johangirod committed:
add comparison

EmileRolley commented:
Trop bien de reprendre ça :pray: 

> Je bloque sur le fait de ne pas permettre les mots-clés « oui » et « non » comme identifier

Tu as un exemple de test qui ne passent pas ?

> Et la prochaine étape : utiliser un [scanner externe](https://tree-sitter.github.io/tree-sitter/creating-parsers#external-scanners) pour l'indentation.

Yes c'est le point bloquant pour l'instant..

EmileRolley commented on package.json:
```diff
@@ -3,14 +3,17 @@
     "parse": "yarn run generate && tree-sitter parse",
-    "test": "yarn run generate && tree-sitter test"
+    "test": "yarn run generate && tree-sitter test",
+    "install": "node-gyp-build",
+    "prebuildify": "prebuildify --napi --strip"
```
Ca sert à quoi tout ça ?

johangirod replied:
Aucune idée c'est venu avec la dernière version de tree-sitter ;)

EmileRolley commented:
About excluding keyword from rule names: https://github.com/tree-sitter/tree-sitter/pull/1635

It seems that for the moment we have to use an external parser to solve this problem...

johangirod committed:
Add indentation support and valeur mecanism parsing

johangirod committed:
wip

johangirod committed:
# Ceci est la combinaison de 5 commits.

# Ceci est le premier message de validation :

feat: add syntaxe for meta fields and mechanisms

lignes ; commençant par ';' seront ignorées, et un message vide
abandonne la validation. ; ; Date :       Thu Apr 18 17:04:55 2024 +0200
effectuées (5 commandes effectuées) : ;    squash 568dc69 wip ;
squash c70d98b wip ; Aucune commande restante. ; Vous êtes en train de
rebaser la branche 'TMP-improve-grammar' sur '32e5388'. ; ;
Modifications qui seront validées : ;	nouveau fichier :
.vscode/launch.json ;	modifié :         .vscode/settings.json ;	nouveau
fichier : .vscode/tasks.json ;	modifié :         grammar.js ;	modifié :
       package.json ;	modifié :         src/grammar.json ;	modifié :
    src/node-types.json ;	modifié :         src/parser.c ;	modifié :
    src/scanner.c ;	modifié :
test/corpus/arithmatic-expression.txt ;	modifié :
test/corpus/mechanisms.txt ;	modifié :         test/corpus/rule.txt

EmileRolley commented on test/corpus/rule_meta.txt:
```diff
@@ -0,0 +1,323 @@
+Meta multiple lines with starting line
+================================================================================
+x: 
+  description: Lorsque la durée d'un contrat de travail est : "inférieure"
```
Est-ce que pour les attributs qui sont connus et utilisés par le moteur (`description`, `titre`, `références`, etc...) ne devraient pas avoir leur propre noeud ? Cela permettra de les différencier des attributs arbitraire propre à chaque projets.

johangirod replied:
Oui, c'est prévu. Je pensais avoir un noeud `custom-meta` pour ces derniers.

EmileRolley replied:
Tu penses que c'est plus intéressant de les regrouper plutôt qu'avoir un noeud pour chaque attribut ?

johangirod replied:
Bonne question. Peut-être qu'il est mieux d'avoir un noeud spécifique pour chaque attribut afin de forcer certains types. Ne pas permettre de paragraphe pour les titres par exemple. Ou bien forcer l'écriture sous forme d'objet pour le noeud référence. En tous les cas c'est facile à modifier.
